File: content/scala-driver/v5.4/source/includes/aggregation.scala
import org.mongodb.scala._
import org.mongodb.scala.model.{ Aggregates, Filters, Accumulators }
import org.mongodb.scala.bson.Document
import com.mongodb.ExplainVerbosity
import org.mongodb.scala.model.Projections
import org.mongodb.scala.model.search._
import org.mongodb.scala.model.search.SearchOptions.searchOptions
import org.mongodb.scala.model.search.SearchPath.fieldPath
import scala.jdk.CollectionConverters._

object Aggregation {

  def main(args: Array[String]): Unit = {
    val mongoClient = MongoClient("<connection string URI>")

    val database: MongoDatabase = mongoClient.getDatabase("sample_restaurants")
    val collection: MongoCollection[Document] = database.getCollection("restaurants")

    // Retrieves documents with a cuisine value of "Bakery", groups them by "borough", and
    // counts each borough's matching documents
    // start-match-group
    val pipeline = Seq(Aggregates.filter(Filters.equal("cuisine", "Bakery")),
                       Aggregates.group("$borough", Accumulators.sum("count", 1))
    )

    collection.aggregate(pipeline)
              .subscribe((doc: Document) => println(doc.toJson()),
                        (e: Throwable) => println(s"There was an error: $e"))
    // end-match-group

    // Performs the same aggregation operation as above but asks MongoDB to explain it
    // start-explain
    val pipelineToExplain = Seq(Aggregates.filter(Filters.equal("cuisine", "Bakery")),
                       Aggregates.group("$borough", Accumulators.sum("count", 1))
    )

    collection.aggregate(pipelineToExplain)
              .explain(ExplainVerbosity.EXECUTION_STATS)
              .subscribe((doc: Document) => println(doc.toJson()),
                        (e: Throwable) => println(s"There was an error: $e"))
    // end-explain

    // start-atlas-search
    val operator = SearchOperator.text(SearchPath.fieldPath("name"), "Salt")
    val options = searchOptions().index("<search index name>")
    
    val pipeline = Seq(Aggregates.search(operator, options),
                       Aggregates.project(Projections.include("name")))

    collection.aggregate(pipeline)
              .subscribe((doc: Document) => println(doc.toJson()),
                        (e: Throwable) => println(s"There was an error: $e"))
    // end-atlas-search

    // start-atlas-helper-methods
    val searchStage = Aggregates.search(
      SearchOperator.compound()
        .must(
          Iterable(
            SearchOperator.in(fieldPath("genres"), List("Comedy")),
            SearchOperator.phrase(fieldPath("fullplot"), "new york"),
            SearchOperator.numberRange(fieldPath("year")).gtLt(1950, 2000),
            SearchOperator.wildcard("Love *", fieldPath("title")),
          ).asJava
        )
    )

    val projectStage = Aggregates.project(
      Projections.include("title", "year", "genres"))

    val aggregatePipelineStages = Seq(searchStage, projectStage)

    collection.aggregate(aggregatePipelineStages)
      .subscribe((doc: Document) => println(doc.toJson()),
        (e: Throwable) => println(s"There was an error: $e"))
    // end-atlas-helper-methods

    Thread.sleep(1000)
    mongoClient.close()
  }
}

File: current/source/includes/aggregation/aggregation.kt
import com.mongodb.ConnectionString
import com.mongodb.MongoClientSettings
import com.mongodb.client.model.Accumulators
import com.mongodb.client.model.Aggregates
import com.mongodb.client.model.Filters
import com.mongodb.kotlin.client.MongoClient
import com.mongodb.client.model.Projections
import com.mongodb.client.model.search.SearchOperator
import com.mongodb.client.model.search.SearchPath.fieldPath
import org.bson.Document

// start-data-class
data class Restaurant(
    val name: String,
    val cuisine: String,
    val borough: String
)
// end-data-class

fun main() {
    val uri = "<connection string URI>"

    val settings = MongoClientSettings.builder()
        .applyConnectionString(ConnectionString(uri))
        .retryWrites(true)
        .build()

    val mongoClient = MongoClient.create(settings)
    val database = mongoClient.getDatabase("sample_restaurants")
    val collection = database.getCollection<Restaurant>("restaurants")

    // start-aggregation-pipeline
    val pipeline = listOf(
        Aggregates.match(Filters.eq(Restaurant::cuisine.name, "Bakery")),
        Aggregates.group("\$borough", Accumulators.sum("count", 1))
    )

    val results = collection.aggregate<Document>(pipeline)

    results.forEach { result ->
        println(result)
    }
    // end-aggregation-pipeline

    // start-aggregation-explain
    print(collection.aggregate(pipeline).explain())
    // end-aggregation-explain

    // start-atlas-searchoperator-helpers
    val searchStage = Aggregates.search(
        SearchOperator.compound()
            .filter(
                listOf(
                    SearchOperator.`in`(fieldPath("genres"), listOf("Comedy")),
                    SearchOperator.phrase(fieldPath("fullplot"), "new york"),
                    SearchOperator.numberRange(fieldPath("year")).gtLt(1950, 2000),
                    SearchOperator.wildcard(fieldPath("title"), "Love *")
                )
            )
    )

    val projectStage = Aggregates.project(
        Projections.include("title", "year", "genres"))

    val pipeline = listOf(searchStage, projectStage)
    val results = collection.aggregate(pipeline)

    results.forEach { result -> println(result) }
    // end-atlas-searchoperator-helpers
}

